README
NAME
    Lexical::Types - Extend the semantics of typed lexicals.

VERSION
    Version 0.13

SYNOPSIS
        { package Str; }

        {
         package My::Types::Str;

         sub new { bless { }, shift }
        }

        use Lexical::Types as => sub { 'My::Types::' . $_[0] => 'new' };

        my Str $x; # $x is now a My::Types::Str object

        {
         package My::Types::Int;

         sub TYPEDSCALAR { bless { }, shift }
        }

        use Lexical::Types;

        use constant Int => 'My::Types::Int';

        my Int $y; # $y is now a My::Types::Int object

DESCRIPTION
    This pragma allows you to hook the execution of typed lexicals
    declarations ("my Str $x") by calling a configurable method in a
    configurable package at each run. In particular, it can be used to
    automatically tie or bless typed lexicals whenever they are initialized.

    Remind that for "perl" to be able to parse "my Str $x", you need :

    *   either the "Str" package to be defined ;

    *   or for "Str" to be a constant sub returning a valid defined package.

    so make sure you follow one of those two strategies to define your
    types.

    This pragma is not implemented with a source filter.

METHODS
  "import"
        use Lexical::Types;
        use Lexical::Types as => $prefix;
        use Lexical::Types as => sub { ... }; # = $mangler

    Magically called when "use Lexical::Types" is encountered. All the
    occurences of "my Str $x" in the current lexical scope will be changed
    to call at each run a given method in a given package. The method and
    package are determined by the parameter 'as' :

    *   If it's left unspecified, the "TYPEDSCALAR" method in the "Str"
        package will be called.

            use Lexical::Types;
            my Str $x; # calls Str->TYPEDSCALAR

    *   If a plain scalar $prefix is passed as the value, the "TYPEDSCALAR"
        method in the "${prefix}::Str" package will be used.

            use Lexical::Types as => 'My::'; # or "as => 'My'"
            my Str $x; # calls My::Str->TYPEDSCALAR

    *   If the value given is a code reference $mangler, it will be called
        at compile-time with arguments 'Str' and 'TYPEDSCALAR' and is
        expected to return :

        *   either an empty list, in which case the current typed lexical
            definition will be skipped (thus it won't be altered to trigger
            a run-time hook) ;

                use Lexical::Types as => sub {
                 return $_[0] =~ /Str/ ? @_ : ()
                };
                my Str $y; # calls Str->TYPEDSCALAR
                my Int $x; # nothing special

        *   or the desired package and method name, in that order (if any of
            those is "undef", the default value will be used instead).

                use Lexical::Types as => sub { 'My', 'new_' . lc($_[0]) };
                my Str $x; # the coderef indicates to call My->new_str

        Note that if the type is a constant, $_[0] will be set to the
        *value* of constant and not to its name.

            use Lexical::Types as => sub { $_[0] => 'new' };
            use constant Str => 'MyStr';
            my Str $x; # calls MyStr->new

        This means in particular that you can't both use constant types and
        redirect several types to different methods of the same package,
        because then you can't distinguish between the original types with
        $_[0].

  "unimport"
        no Lexical::Types;

    Magically called when "no Lexical::Types" is encountered. Turns the
    pragma off.

RUN-TIME INITIALIZER METHOD
    The initializer method receives an alias to the pad slot of the
    initialized lexical in $_[1] and the original type name in $_[2]. You
    can either edit $_[1] in place, in which case you should return an empty
    list, or return a new scalar that will be copied into the pad slot.

        use Lexical::Types as => 'My';

        my Str $x;

        ...

        sub My::Str::TYPEDSCALAR {
         # $_[1] is an alias to $x, and $_[2] is 'Str'
         ...
        }

INTEGRATION
    You can integrate Lexical::Types in your module so that using it will
    provide types to your users without asking them to load either
    Lexical::Types or the type classes manually.

        package MyTypes;

        BEGIN { require Lexical::Types; }

        sub import {
         eval 'package Str; package Int'; # The types you want to support
         Lexical::Types->import(
          as => sub { __PACKAGE__, 'new_' . lc($_[0]) }
         );
        }

        sub unimport {
         Lexical::Types->unimport;
        }

        sub new_str { ... }

        sub new_int { ... }

    If you prefer to use constants rather than creating empty packages, you
    can replace the previous example with something like this :

        package MyTypes;

        BEGIN { require Lexical::Types; }

        sub import {
         my $pkg = caller;
         for (qw<Str Int>) {
          my $type = __PACKAGE__ . '::' . $_;
          no strict 'refs';
          no warnings 'redefine';
          *{$pkg.'::'.$_} = eval "sub () { '$type' }";
         }
         Lexical::Types->import(
          as => sub { $_[0] => 'new' }
         );
        }

        sub unimport {
         Lexical::Types->unimport;
        }

        package MyTypes::Str;

        sub new { ... }

        package MyTypes::Int;

        sub new { ... }

CONSTANTS
  "LT_THREADSAFE"
    True iff the module could have been built with thread-safety features
    enabled.

  "LT_FORKSAFE"
    True iff this module could have been built with fork-safety features
    enabled. This will always be true except on Windows where it's false for
    perl 5.10.0 and below .

CAVEATS
    Using this pragma will cause a slight global slowdown of any subsequent
    compilation phase that happens anywere in your code - even outside of
    the scope of use of "use Lexical::Types" - which may become noticeable
    if you rely heavily on numerous calls to "eval STRING".

    The restrictions on the type (being either a defined package name or a
    constant) apply even if you use the 'as' option to redirect to another
    package, and are unlikely to find a workaround as this happens deep
    inside the lexer - far from the reach of an extension.

    Only one mangler or prefix can be in use at the same time in a given
    scope.

    Typed lexicals declarations that appear in code "eval"'d during the
    global destruction phase of a spawned thread or pseudo-fork (the
    processes used internally for the "fork" emulation on Windows) are
    ignored.

    The implementation was tweaked to work around several limitations of
    vanilla "perl" pragmas : it's thread safe, and doesn't suffer from a
    "perl 5.8.x-5.10.0" bug that causes all pragmas to propagate into
    "require"d scopes.

    With 5.8 perls, the pragma does not propagate into "eval STRING". This
    is due to a shortcoming in the way perl handles the hints hash, which is
    addressed in perl 5.10.

DEPENDENCIES
    perl 5.8.3.

    A C compiler. This module may happen to build with a C++ compiler as
    well, but don't rely on it, as no guarantee is made in this regard.

    XSLoader (standard since perl 5.6.0).

SEE ALSO
    fields.

    Attribute::Handlers.

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-lexical-types at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Lexical-Types>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Lexical::Types

    Tests code coverage report is available at
    <http://www.profvince.com/perl/cover/Lexical-Types>.

ACKNOWLEDGEMENTS
    Inspired by Ricardo Signes.

    Thanks Florian Ragwitz for suggesting the use of constants for types.

COPYRIGHT & LICENSE
    Copyright 2009,2010,2011,2012,2013,2014 Vincent Pit, all rights
    reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Lexical-Types

0.13    2014-10-04 23:25 UTC
        + Add : Support for the PERL_OP_PARENT optional feature introduced in
                perl 5.21.2.
        + Doc : The CAVEATS section now warns about the global slowdown during
                compilation caused by this pragma.
        + Fix : [RT #86112] : Doesn't hook new PADRANGE op in Perl 5.18.
                The PADRANGE optimization is not fully supported.
                Thanks Dagfinn Ilmari Mannsåker for reporting.
        + Fix : Segfaults in eval in an END block of a Win32 pseudo-fork.
        + Fix : Segfaults during global destruction of a thread or a
                pseudo-fork.
        + Fix : Check functions are now replaced and restored in a thread-safe
                manner, either by using the wrap_op_checker() function from perl
                when it is available (starting from perl 5.16) or by taking the
                OP_REFCNT mutex on older perls.
        + Tst : Author tests are no longer bundled with this distribution.
                They are only made available to authors in the git repository.
        + Tst : Test failures of t/81-threads-teardown.t on Cygwin and Android
                should have been addressed.
        + Tst : Threads tests will not fail anymore if resources constraints
                prevent the system from creating all the required threads.
        + Upd : Metadata overhaul.

0.12    2011-08-24 16:30 UTC
        + Fix : The pragma no longer vivifies the "Lexical::Types" entry in the
                hints hash %^H on perl 5.8.

0.11    2011-02-26 22:00 UTC
        + Fix : [RT #66164] : Lexical::Types hangs with PPI.
                This was actually a regression introduced together with the new
                peephole optimizer strategy, and that caused the pragma to hang
                on constructs like "for (;;) { ... }".
                Thanks Chisel Wright for reminding me about this issue.

0.10    2011-01-03 20:35 UTC
        + Add : The new constant LT_FORKSAFE can be tested to know whether the
                module will behave nicely when fork()ing. It's currently always
                true except on Windows where you need perl 5.10.1 for it to be
                true.
        + Chg : perl 5.8.3 is now required (instead of 5.8.0).
        + Fix : Scope leaks under perl 5.8-5.10.0.
        + Fix : Segmentation faults and misbehaviours in threaded applications.
        + Fix : Compatibility with perl 5.13.1 and higher.
        + Fix : Broken linkage on Windows with gcc 3.4, which appears in
                particular when using ActivePerl's default compiler suite.
                For those setups, the Lexical::Types shared library will now
                be linked against the perl dll directly (instead of the import
                library).
        + Tst : Threads tests are now only run on perl 5.13.4 and higher.
                They could segfault randomly because of what seems to be an
                internal bug of Perl, which has been addressed in 5.13.4.
                There is also an environment variable that allows you to
                forcefully run those tests, but it should be set only for
                author testing and not for end users.

0.09    2010-01-03 00:00 UTC
        + Fix : Building and testing with blead.
        + Fix : Unbalanced scopes when skipping a typed declaration.
        + Fix : Segfaults when Lexical::Types is loaded for the first time from
                inside a thread.
        + Fix : Leaks of memory associated with the root interpreter.
        + Fix : Work around Kwalitee test misfailures.
        + Opt : Less memory will be used for non-threaded perls version 5.10.0
                and below, and for threaded perls from version 5.10.1.

0.08    2009-07-04 19:35 UTC
        + Fix : Don't leak the old op info when a pointer table entry is reused.
        + Fix : Possibly missed constructs with eval STRING called in a thread.

0.07    2009-07-01 08:45 UTC
        + Chg : perl 5.8.x on MSWin32 is no longer considered as thread safe.
        + Fix : Work around a bug in perl 5.10.0 and lower that cause hints to
                propagate into required files.
        + Fix : Clean up map entries associated to uncatched OPs.

0.06    2009-05-01 19:00 UTC
        + Fix : The pragma now properly propagates into threaded eval STRING,
                effectively making the module thread safe. The new LT_THREADSAFE
                constant can be used to know whether the module could have been
                built with thread safety features enabled.

0.05    2009-04-28 21:40 UTC
        + Fix : Building with PERL_IMPLICIT_SYS set (especially on Win32).
        + Tst : Dieing in callbacks.

0.04    2009-03-07 15:45 UTC
        + Chg : Some implementation of pointer table is now used internally.
                This should make thread safety more robust.
        + Doc : Cleanups.
        + Tst : Using Lexical::Types inside the 'as' callback.
        + Tst : Thread safety.

0.03    2009-03-05 21:15 UTC
        + Doc : Discuss about using constants for types.
        + Fix : The PL_ppaddr[OP_PADSV] localization logic was refined so that
                it is kinder to other modules that replace it.
        + Tst : "my Str ($x, $y)" and "for my Str $x ()" constructs
        + Tst : uvar magic, magical tags.

0.02    2009-02-25 16:10 UTC
        + Add : Returning an empty list from the mangler skips the wrapping of
                the current typed lexical declaration.
        + Chg : The package and method names passed to the callbacks are now all
                read-only.

0.01    2009-02-24 23:20 UTC
        First version, released on an unsuspecting world.



t/lib/Lexical/Types/TestThreads.pm
package Lexical::Types::TestThreads;

use strict;
use warnings;

use Config qw<%Config>;

use VPIT::TestHelpers;

sub import {
 shift;

 require Lexical::Types;

 skip_all 'This Lexical::Types isn\'t thread safe'
                                         unless Lexical::Types::LT_THREADSAFE();

 my $force = $ENV{PERL_LEXICAL_TYPES_TEST_THREADS} ? 1 : !1;
 skip_all 'This perl wasn\'t built to support threads'
                                                    unless $Config{useithreads};
 skip_all 'perl 5.13.4 required to test thread safety'
                                             unless $force or "$]" >= 5.013_004;

 load_or_skip_all('threads', $force ? '0' : '1.67', [ ]);

 my %exports = (
  spawn => \&spawn,
 );

 my $pkg = caller;
 while (my ($name, $code) = each %exports) {
  no strict 'refs';
  *{$pkg.'::'.$name} = $code;
 }
}

sub spawn {
 local $@;
 my @diag;
 my $thread = eval {
  local $SIG{__WARN__} = sub { push @diag, "Thread creation warning: @_" };
  threads->create(@_);
 };
 push @diag, "Thread creation error: $@" if $@;
 if (@diag) {
  require Test::Leaner;
  Test::Leaner::diag($_) for @diag;
 }
 return $thread ? $thread : ();
}

1;


t/lib/Lexical/Types/TestRequired2.pm
package Lexical::Types::TestRequired2;

use Lexical::Types;

BEGIN {
 delete $INC{'Lexical/Types/TestRequired1.pm'};
}

use lib 't/lib';
use Lexical::Types::TestRequired1;

my Int $x;
Test::More::is($x, 't/lib/Lexical/Types/TestRequired2.pm:' . (__LINE__-1), 'pragma in use in require');

eval q!
 my Int $y;
 my $desc = 'pragma in use in eval in require';
 if ("$]" <  5.009005) {
  Test::More::is($y, undef, $desc);
 } else {
  Test::More::like($y, qr/^\(eval +\d+\):2$/, $desc);
 }
!;

1;


reap.h
/* This file is part of the Lexical::Types Perl module.
 * See http://search.cpan.org/dist/Lexical-Types/ */

/* This header provides a specialized version of Scope::Upper::reap that can be
 * called directly from XS.
 * See http://search.cpan.org/dist/Scope-Upper/ for details. */

#ifndef REAP_H
#define REAP_H 1

#define REAP_DESTRUCTOR_SIZE 3

typedef struct {
 I32    depth;
 I32   *origin;
 void (*cb)(pTHX_ void *);
 void  *ud;
 char  *dummy;
} reap_ud;

STATIC void reap_pop(pTHX_ void *);

STATIC void reap_pop(pTHX_ void *ud_) {
 reap_ud *ud = ud_;
 I32 depth, *origin, mark, base;

 depth  = ud->depth;
 origin = ud->origin;
 mark   = origin[depth];
 base   = origin[depth - 1];

 if (base < mark) {
  PL_savestack_ix = mark;
  leave_scope(base);
 }
 PL_savestack_ix = base;

 if ((ud->depth = --depth) > 0) {
  SAVEDESTRUCTOR_X(reap_pop, ud);
 } else {
  void (*cb)(pTHX_ void *) = ud->cb;
  void  *cb_ud             = ud->ud;

  PerlMemShared_free(ud->origin);
  PerlMemShared_free(ud);

  SAVEDESTRUCTOR_X(cb, cb_ud);
 }
}

STATIC void reap(pTHX_ I32 depth, void (*cb)(pTHX_ void *), void *cb_ud) {
#define reap(D, CB, UD) reap(aTHX_ (D), (CB), (UD))
 reap_ud *ud;
 I32 i;

 if (depth > PL_scopestack_ix)
  depth = PL_scopestack_ix;

 ud         = PerlMemShared_malloc(sizeof *ud);
 ud->depth  = depth;
 ud->origin = PerlMemShared_malloc((depth + 1) * sizeof *ud->origin);
 ud->cb     = cb;
 ud->ud     = cb_ud;
 ud->dummy  = NULL;

 for (i = depth; i >= 1; --i) {
  I32 j = PL_scopestack_ix - i;
  ud->origin[depth - i] = PL_scopestack[j];
  PL_scopestack[j] += REAP_DESTRUCTOR_SIZE;
 }
 ud->origin[depth] = PL_savestack_ix;

 while (PL_savestack_ix + REAP_DESTRUCTOR_SIZE
                                       <= PL_scopestack[PL_scopestack_ix - 1]) {
  save_pptr(&ud->dummy);
 }

 SAVEDESTRUCTOR_X(reap_pop, ud);
}

#endif /* REAP_H */


ptable.h
/* This file is part of the Lexical::Types Perl module.
 * See http://search.cpan.org/dist/Lexical-Types/ */

/* This is a pointer table implementation essentially copied from the ptr_table
 * implementation in perl's sv.c, except that it has been modified to use memory
 * shared across threads.
 * Copyright goes to the original authors, bug reports to me. */

/* This header is designed to be included several times with different
 * definitions for PTABLE_NAME and PTABLE_VAL_FREE(). */

#undef VOID2
#ifdef __cplusplus
# define VOID2(T, P) static_cast<T>(P)
#else
# define VOID2(T, P) (P)
#endif

#undef pPTBLMS
#undef pPTBLMS_
#undef aPTBLMS
#undef aPTBLMS_

/* Context for PerlMemShared_* functions */

#ifdef PERL_IMPLICIT_SYS
# define pPTBLMS  pTHX
# define pPTBLMS_ pTHX_
# define aPTBLMS  aTHX
# define aPTBLMS_ aTHX_
#else
# define pPTBLMS  void
# define pPTBLMS_
# define aPTBLMS
# define aPTBLMS_
#endif

#ifndef pPTBL
# define pPTBL  pPTBLMS
#endif
#ifndef pPTBL_
# define pPTBL_ pPTBLMS_
#endif
#ifndef aPTBL
# define aPTBL  aPTBLMS
#endif
#ifndef aPTBL_
# define aPTBL_ aPTBLMS_
#endif

#ifndef PTABLE_NAME
# define PTABLE_NAME ptable
#endif

#ifndef PTABLE_VAL_FREE
# define PTABLE_VAL_FREE(V)
#endif

#ifndef PTABLE_JOIN
# define PTABLE_PASTE(A, B) A ## B
# define PTABLE_JOIN(A, B)  PTABLE_PASTE(A, B)
#endif

#ifndef PTABLE_PREFIX
# define PTABLE_PREFIX(X) PTABLE_JOIN(PTABLE_NAME, X)
#endif

#ifndef ptable_ent
typedef struct ptable_ent {
 struct ptable_ent *next;
 const void *       key;
 void *             val;
} ptable_ent;
#define ptable_ent ptable_ent
#endif /* !ptable_ent */

#ifndef ptable
typedef struct ptable {
 ptable_ent **ary;
 size_t       max;
 size_t       items;
} ptable;
#define ptable ptable
#endif /* !ptable */

#ifndef ptable_new
STATIC ptable *ptable_new(pPTBLMS) {
#define ptable_new() ptable_new(aPTBLMS)
 ptable *t = VOID2(ptable *, PerlMemShared_malloc(sizeof *t));
 t->max    = 15;
 t->items  = 0;
 t->ary    = VOID2(ptable_ent **,
                              PerlMemShared_calloc(t->max + 1, sizeof *t->ary));
 return t;
}
#endif /* !ptable_new */

#ifndef PTABLE_HASH
# define PTABLE_HASH(ptr) \
     ((PTR2UV(ptr) >> 3) ^ (PTR2UV(ptr) >> (3 + 7)) ^ (PTR2UV(ptr) >> (3 + 17)))
#endif

#ifndef ptable_find
STATIC ptable_ent *ptable_find(const ptable * const t, const void * const key) {
#define ptable_find ptable_find
 ptable_ent *ent;
 const UV hash = PTABLE_HASH(key);

 ent = t->ary[hash & t->max];
 for (; ent; ent = ent->next) {
  if (ent->key == key)
   return ent;
 }

 return NULL;
}
#endif /* !ptable_find */

#ifndef ptable_fetch
STATIC void *ptable_fetch(const ptable * const t, const void * const key) {
#define ptable_fetch ptable_fetch
 const ptable_ent *const ent = ptable_find(t, key);

 return ent ? ent->val : NULL;
}
#endif /* !ptable_fetch */

#ifndef ptable_split
STATIC void ptable_split(pPTBLMS_ ptable * const t) {
#define ptable_split(T) ptable_split(aPTBLMS_ (T))
 ptable_ent **ary = t->ary;
 const size_t oldsize = t->max + 1;
 size_t newsize = oldsize * 2;
 size_t i;

 ary = VOID2(ptable_ent **, PerlMemShared_realloc(ary, newsize * sizeof(*ary)));
 Zero(&ary[oldsize], newsize - oldsize, sizeof(*ary));
 t->max = --newsize;
 t->ary = ary;

 for (i = 0; i < oldsize; i++, ary++) {
  ptable_ent **curentp, **entp, *ent;
  if (!*ary)
   continue;
  curentp = ary + oldsize;
  for (entp = ary, ent = *ary; ent; ent = *entp) {
   if ((newsize & PTABLE_HASH(ent->key)) != i) {
    *entp     = ent->next;
    ent->next = *curentp;
    *curentp  = ent;
    continue;
   } else
    entp = &ent->next;
  }
 }
}
#endif /* !ptable_split */

STATIC void PTABLE_PREFIX(_store)(pPTBL_ ptable * const t, const void * const key, void * const val) {
 ptable_ent *ent = ptable_find(t, key);

 if (ent) {
  void *oldval = ent->val;
  PTABLE_VAL_FREE(oldval);
  ent->val = val;
 } else if (val) {
  const size_t i = PTABLE_HASH(key) & t->max;
  ent = VOID2(ptable_ent *, PerlMemShared_malloc(sizeof *ent));
  ent->key  = key;
  ent->val  = val;
  ent->next = t->ary[i];
  t->ary[i] = ent;
  t->items++;
  if (ent->next && t->items > t->max)
   ptable_split(t);
 }
}

STATIC void PTABLE_PREFIX(_delete)(pPTBL_ ptable * const t, const void * const key) {
 ptable_ent *prev, *ent;
 const size_t i = PTABLE_HASH(key) & t->max;

 prev = NULL;
 ent  = t->ary[i];
 for (; ent; prev = ent, ent = ent->next) {
  if (ent->key == key)
   break;
 }

 if (ent) {
  if (prev)
   prev->next = ent->next;
  else
   t->ary[i]  = ent->next;
  PTABLE_VAL_FREE(ent->val);
  PerlMemShared_free(ent);
 }
}

#ifndef ptable_walk
STATIC void ptable_walk(pTHX_ ptable * const t, void (*cb)(pTHX_ ptable_ent *ent, void *userdata), void *userdata) {
#define ptable_walk(T, CB, UD) ptable_walk(aTHX_ (T), (CB), (UD))
 if (t && t->items) {
  register ptable_ent ** const array = t->ary;
  size_t i = t->max;
  do {
   ptable_ent *entry;
   for (entry = array[i]; entry; entry = entry->next)
    if (entry->val)
     cb(aTHX_ entry, userdata);
  } while (i--);
 }
}
#endif /* !ptable_walk */

STATIC void PTABLE_PREFIX(_clear)(pPTBL_ ptable * const t) {
 if (t && t->items) {
  register ptable_ent ** const array = t->ary;
  size_t i = t->max;

  do {
   ptable_ent *entry = array[i];
   while (entry) {
    ptable_ent * const oentry = entry;
    void *val = oentry->val;
    entry = entry->next;
    PTABLE_VAL_FREE(val);
    PerlMemShared_free(oentry);
   }
   array[i] = NULL;
  } while (i--);

  t->items = 0;
 }
}

STATIC void PTABLE_PREFIX(_free)(pPTBL_ ptable * const t) {
 if (!t)
  return;
 PTABLE_PREFIX(_clear)(aPTBL_ t);
 PerlMemShared_free(t->ary);
 PerlMemShared_free(t);
}

#undef pPTBL
#undef pPTBL_
#undef aPTBL
#undef aPTBL_

#undef PTABLE_NAME
#undef PTABLE_VAL_FREE


lib/Lexical/Types.pm
package Lexical::Types;

use 5.008_003;

use strict;
use warnings;

=head1 NAME

Lexical::Types - Extend the semantics of typed lexicals.

=head1 VERSION

Version 0.13

=cut

our $VERSION;
BEGIN {
 $VERSION = '0.13';
}

=head1 SYNOPSIS

    { package Str; }

    {
     package My::Types::Str;

     sub new { bless { }, shift }
    }

    use Lexical::Types as => sub { 'My::Types::' . $_[0] => 'new' };

    my Str $x; # $x is now a My::Types::Str object

    {
     package My::Types::Int;

     sub TYPEDSCALAR { bless { }, shift }
    }

    use Lexical::Types;

    use constant Int => 'My::Types::Int';

    my Int $y; # $y is now a My::Types::Int object

=head1 DESCRIPTION

This pragma allows you to hook the execution of typed lexicals declarations (C<my Str $x>) by calling a configurable method in a configurable package at each run.
In particular, it can be used to automatically tie or bless typed lexicals whenever they are initialized.

Remind that for C<perl> to be able to parse C<my Str $x>, you need :

=over 4

=item *

either the C<Str> package to be defined ;

=item *

or for C<Str> to be a constant sub returning a valid defined package.

=back

so make sure you follow one of those two strategies to define your types.

This pragma is B<not> implemented with a source filter.

=cut

BEGIN {
 require XSLoader;
 XSLoader::load(__PACKAGE__, $VERSION);
}

=head1 METHODS

=head2 C<import>

    use Lexical::Types;
    use Lexical::Types as => $prefix;
    use Lexical::Types as => sub { ... }; # = $mangler

Magically called when C<use Lexical::Types> is encountered.
All the occurences of C<my Str $x> in the current lexical scope will be changed to call at each run a given method in a given package.
The method and package are determined by the parameter C<'as'> :

=over 4

=item *

If it's left unspecified, the C<TYPEDSCALAR> method in the C<Str> package will be called.

    use Lexical::Types;
    my Str $x; # calls Str->TYPEDSCALAR

=item *

If a plain scalar C<$prefix> is passed as the value, the C<TYPEDSCALAR> method in the C<${prefix}::Str> package will be used.

    use Lexical::Types as => 'My::'; # or "as => 'My'"
    my Str $x; # calls My::Str->TYPEDSCALAR

=item *

If the value given is a code reference C<$mangler>, it will be called at compile-time with arguments C<'Str'> and C<'TYPEDSCALAR'> and is expected to return :

=over 4

=item *

either an empty list, in which case the current typed lexical definition will be skipped (thus it won't be altered to trigger a run-time hook) ;

    use Lexical::Types as => sub {
     return $_[0] =~ /Str/ ? @_ : ()
    };
    my Str $y; # calls Str->TYPEDSCALAR
    my Int $x; # nothing special

=item *

or the desired package and method name, in that order (if any of those is C<undef>, the default value will be used instead).

    use Lexical::Types as => sub { 'My', 'new_' . lc($_[0]) };
    my Str $x; # the coderef indicates to call My->new_str

=back

Note that if the type is a constant, C<$_[0]> will be set to the I<value> of constant and not to its name.

    use Lexical::Types as => sub { $_[0] => 'new' };
    use constant Str => 'MyStr';
    my Str $x; # calls MyStr->new

This means in particular that you can't both use constant types and redirect several types to different methods of the same package, because then you can't distinguish between the original types with C<$_[0]>.

=back

=cut

sub import {
 shift;
 my %args = @_;

 my $hint;

 my $as = delete $args{'as'};
 if ($as) {
  my $r = ref $as;
  if ($r eq 'CODE') {
   $hint = _tag($as);
  } elsif (!$r) {
   $as .= '::' if $as !~ /::$/;
   $hint = _tag(sub { $as . $_[0] });
  } else {
   require Carp;
   Carp::croak("Invalid $r reference for 'as'");
  }
 } else {
  $hint = _tag(sub { @_ });
 }

 $^H |= 0x020000;
 # Yes, we store a coderef inside the hints hash, but that's just for compile
 # time.
 $^H{+(__PACKAGE__)} = $hint;
}

=head2 C<unimport>

    no Lexical::Types;

Magically called when C<no Lexical::Types> is encountered.
Turns the pragma off.

=cut

sub unimport {
 $^H{+(__PACKAGE__)} = undef;
}

=head1 RUN-TIME INITIALIZER METHOD

The initializer method receives an alias to the pad slot of the initialized lexical in C<$_[1]> and the original type name in C<$_[2]>.
You can either edit C<$_[1]> in place, in which case you should return an empty list, or return a new scalar that will be copied into the pad slot.

    use Lexical::Types as => 'My';

    my Str $x;

    ...

    sub My::Str::TYPEDSCALAR {
     # $_[1] is an alias to $x, and $_[2] is 'Str'
     ...
    }

=head1 INTEGRATION

You can integrate L<Lexical::Types> in your module so that using it will provide types to your users without asking them to load either L<Lexical::Types> or the type classes manually.

    package MyTypes;

    BEGIN { require Lexical::Types; }

    sub import {
     eval 'package Str; package Int'; # The types you want to support
     Lexical::Types->import(
      as => sub { __PACKAGE__, 'new_' . lc($_[0]) }
     );
    }

    sub unimport {
     Lexical::Types->unimport;
    }

    sub new_str { ... }

    sub new_int { ... }

If you prefer to use constants rather than creating empty packages, you can replace the previous example with something like this :

    package MyTypes;

    BEGIN { require Lexical::Types; }

    sub import {
     my $pkg = caller;
     for (qw<Str Int>) {
      my $type = __PACKAGE__ . '::' . $_;
      no strict 'refs';
      no warnings 'redefine';
      *{$pkg.'::'.$_} = eval "sub () { '$type' }";
     }
     Lexical::Types->import(
      as => sub { $_[0] => 'new' }
     );
    }

    sub unimport {
     Lexical::Types->unimport;
    }

    package MyTypes::Str;

    sub new { ... }

    package MyTypes::Int;

    sub new { ... }

=head1 CONSTANTS

=head2 C<LT_THREADSAFE>

True iff the module could have been built with thread-safety features enabled.

=head2 C<LT_FORKSAFE>

True iff this module could have been built with fork-safety features enabled.
This will always be true except on Windows where it's false for perl 5.10.0 and below .

=head1 CAVEATS

Using this pragma will cause a slight global slowdown of any subsequent compilation phase that happens anywere in your code - even outside of the scope of use of C<use Lexical::Types> - which may become noticeable if you rely heavily on numerous calls to C<eval STRING>.

The restrictions on the type (being either a defined package name or a constant) apply even if you use the C<'as'> option to redirect to another package, and are unlikely to find a workaround as this happens deep inside the lexer - far from the reach of an extension.

Only one mangler or prefix can be in use at the same time in a given scope.

Typed lexicals declarations that appear in code C<eval>'d during the global destruction phase of a spawned thread or pseudo-fork (the processes used internally for the C<fork> emulation on Windows) are ignored.

The implementation was tweaked to work around several limitations of vanilla C<perl> pragmas : it's thread safe, and doesn't suffer from a C<perl 5.8.x-5.10.0> bug that causes all pragmas to propagate into C<require>d scopes.

With 5.8 perls, the pragma does not propagate into C<eval STRING>.
This is due to a shortcoming in the way perl handles the hints hash, which is addressed in perl 5.10.

=head1 DEPENDENCIES

L<perl> 5.8.3.

A C compiler.
This module may happen to build with a C++ compiler as well, but don't rely on it, as no guarantee is made in this regard.

L<XSLoader> (standard since perl 5.6.0).

=head1 SEE ALSO

L<fields>.

L<Attribute::Handlers>.

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

You can contact me by mail or on C<irc.perl.org> (vincent).

=head1 BUGS

Please report any bugs or feature requests to C<bug-lexical-types at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Lexical-Types>.
I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Lexical::Types

Tests code coverage report is available at L<http://www.profvince.com/perl/cover/Lexical-Types>.

=head1 ACKNOWLEDGEMENTS

Inspired by Ricardo Signes.

Thanks Florian Ragwitz for suggesting the use of constants for types.

=head1 COPYRIGHT & LICENSE

Copyright 2009,2010,2011,2012,2013,2014 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1; # End of Lexical::Types
